### PR TITLE
[common] add a device_memset_async for device_buffer

### PR DESCRIPTION
**Pull request by ahehn-nv**
* [common] device_buffer added initialize() method
* [cudaaligner] AlignerGlobal: initialize device buffers

### CODE REVIEW

File: common/utils/include/claragenomics/utils/device_buffer.cuh [hunk 1]
```diff
@@ -68,6 +68,11 @@ public:
         cudaFree(data_);
     }
 
+    void initialize(cudaStream_t stream)
```
Comment (mimaric) [inline common/utils/include/claragenomics/utils/device_buffer.cuh]: Passing a stream should be enough to indicate that this is an async operation, but it feels like `initialize_async` would be it even clearer.

Comment (ahehn-nv) [reply common/utils/include/claragenomics/utils/device_buffer.cuh]: yes, makes sense. I changed it from `initialize(stream)` to `initialize_async(stream)`. Thanks!

**Comment by mimaric** [comment]
Now that I thought a bit more about it... How about adding a `bool initialize = false` (or an internal enum class) argument to the constructor? Initialization is tightly related to construction and if someone wants to set the array to zero after already doing something with the array they should do it on their own, no need for the class to provide a method for it

**Comment by ahehn-nv** [comment]
I thought about that, but decided against it, because it would make the semantics less clear in my opinion. Right now, the `device_buffer` just provides you bare device memory - period. You have to take care of initializing it (either manually or by calling `initialize_async()`).
If you add an `initialize` argument to the constructor it will be less clear if you can expect the buffer to be initialized or not (at least if you provide a default value), and the meaning of passing the stream to the constructor is not obvious.
I considered `device_buffer buf(size, async_initializer(stream))` for a moment, but `buf.initialize_async(stream)` seemed the easier solution for what I needed.

Maybe the name `initialize` is not ideal? `memset_async(value,stream)` would be more general and maybe a bit clearer...

**Comment by mimaric** [comment]
Yes, I also thought about the ugliness of such constructor...
The more I think about it the more I wonder if we need something like this in `device_buffer` at all. I think we should keep it all simple as possible. If we would add something like this why not add `copy_data_to` and `copy_data_from` as well?
@tijyojwad @akkamesh @vellamike What do you think?

**Comment by ahehn-nv** [comment]
I just wanted to spare myself the lengthy ugly
```
cudaMemset(sequences_d_.data(), 0, sequences_d_.size()*sizeof(char), stream);
cudaMemset(sequence_lengths_d_.data(), 0, sequence_lengths_d_.size()*sizeof(int32_t), stream);
cudaMemset(results_d_.data(), 0, results_d_.size()*sizeof(int8_t), stream);
cudaMemset(result_lengths_d_.data(), 0, result_lengths_d_.size()*sizeof(int32_t), stream);
```
But good point with `copy_data_to/from`...

**Comment by mimaric** [comment]
I'm rather undecided on this one... Let's hear some more opinions...

**Comment by ahehn-nv** [comment]
Yes, me too. 
Yet another possibilty: I could also just add a
```
template <typename T>
device_memset_async(device_buffer<T>& buf, int32_t byte, cudaStream_t stream)
{
    cudaMemsetAsync(buf.data(), byte, sizeof(T)*buf.size(), stream);
}
```
either publicly to the `device_buffer.cuh` to decouple it from the class a bit, or just as utility to the `aligner_global.cpp`. But let's wait for more opinions...

**Comment by tijyojwad** [comment]
@ahehn-nv @mimaric I think it should be part of the `device_buffer` class itself, with an API like `memset_async(byte, stream)`. As for the other functions, it indeed would be nice to have the `copy_to` and `copy_from` APIs too, but I want to be wary of effectively re-writing the thrust `device_vector` APIs which we want to move to soon :)

**Comment by mimaric** [comment]
I would say there are three options:
1) Do nothing
   + keeps `device_buffer` lean
   + `CGA_CU_CHECK_ERR` always reports the line on which memset was done
   - Much typing, hard to read, easy to make mistake
2) Standalone function
   + keeps `device_buffer` lean
   + easy to use
   - `CGA_CU_CHECK_ERR` reports the like withing that function, not where memset was actually done
3) Add method to `device_buffer`
   + easy to use
   +- `device_buffer` interface grows
   - `CGA_CU_CHECK_ERR` reports the like withing that method, not where memset was actually done

I do agree with @tijyojwad that we might be reimplementing `thrust::device_vector`

**Comment by ahehn-nv** [comment]
I slept a night on it and I disagree with @tijyojwad .
I think the `device_buffer` should remain the way it is - minimal - just providing device memory. memset and copies have nothing to do with this functionality. These operations do not manipulate the internal state of the object (the pointer and size), so there is no need to make them methods.

However, I would like to add the previously mentioned convenience function to `device_buffer.cuh`, which uses the existing interface of `device_buffer`:
```
template <typename T>
device_memset_async(device_buffer<T>& buf, int32_t byte, cudaStream_t stream)
{
    cudaMemsetAsync(buf.data(), byte, sizeof(T)*buf.size(), stream);
}
```
It mimics the underlying CUDA call and if we move away from `device_buffer`, this function can also easily be adapted/overloaded.
Any thoughts?

**Comment by ahehn-nv** [comment]
@mimaric and my comments crossed. So, in my comment I propose to go with @mimaric's option 2.
The `CGA_CU_CHECK_ERR` argument is not very helpful in my opinion. If you follow this through it means you should write spaghetti code, doesn't it? I would not expect `CGA_CU_CHEC_ERR` to give you the full error information. If something goes sideways, you should take a debugger to have a closer look and obtain a stacktrace there (with asynchronous execution it might not that helpful anyways).

**Comment by mimaric** [comment]
I agree. Let's go for a standalone function!

**Comment by ahehn-nv** [comment]
I pushed my new proposal. Please review.

**Comment by ahehn-nv** [comment]
@tijyojwad do you have further thoughts on this?

**Comment by tijyojwad** [comment]
@ahehn-nv I'm on board with keeping the class minimal too. Go ahead and merge

**Pull request merged**